src/transforms/sample/transform.rs

```diff
@@ -12,24 +12,86 @@ use crate::{
 use vector_lib::lookup::lookup_v2::OptionalValuePath;
 use vector_lib::lookup::OwnedTargetPath;
 
+/// Exists only for backwards compatability purposes so that the value of sample_rate_key is
+/// consistent after the internal implementation of the Sample class was modified to work in terms
+/// of percentages
+#[derive(Clone, Debug)]
+pub enum SampleMode {
+    Rate(u64, HashMap<Option<String>, u64>),
+    Ratio(f64, HashMap<Option<String>, f64>),
+}
+
+impl SampleMode {
+    pub fn new_rate(rate: u64) -> Self {
+        Self::Rate(rate, HashMap::default())
+    }
+
+    pub fn new_ratio(ratio: f64) -> Self {
+        Self::Ratio(ratio, HashMap::default())
+    }
+
+    fn increment(&mut self, key: &Option<String>) -> bool {
+        match self {
+            Self::Rate(rate, counter) => {
+                let counter_value = counter.entry(key.clone()).or_default();
+                let increment: u64 = *counter_value + 1;
+                *counter_value = increment;
+                increment % *rate == 0
+            }
+            Self::Ratio(ratio, counter) => {
+                let counter_value = counter.entry(key.clone()).or_insert(*ratio);
```

Comment:
nit: My suggestion had be to set this default at `1.0 - ratio` so as to guarantee the first event is emitted. If we don't care about that guarantee `.or_default()` works here as well.